graphql_compiler/compiler/ir_lowering_sql/__init__.py

```diff
@@ -0,0 +1,264 @@
+
+
+def get_block_index_to_location_map(ir_blocks):
+    """Associate each IR block with it's corresponding location, by index."""
+    block_to_location = {}
+    current_block_ixs = []
+    for num, ir_block in enumerate(ir_blocks):
+        current_block_ixs.append(num)
+        if isinstance(ir_block, blocks.MarkLocation):
+            for ix in current_block_ixs:
+                block_to_location[ix] = ir_block.location
+            current_block_ixs = []
+    return block_to_location
+
+
+def lower_construct_result(construct_result):
+    """Lower the ConstructResult IR Block."""
+    def visitor_fn(expression):
+        """Rewrite output fields in optional scopes to their if_true branch.
+
+        The SQL backend does not require special handling to get the correct semantics for such
+        output fields.
+        """
+        if not isinstance(expression, expressions.TernaryConditional):
+            return expression
+        return expression.if_true
+    return construct_result.visit_and_update_expressions(visitor_fn)
+
+
+def lower_unary_transformations(ir_blocks):
+    """Raise errors if any unary transformation block encountered."""
+    def visitor_fn(expression):
+        """Raise error if current expression is a UnaryTransformation."""
+        if not isinstance(expression, expressions.UnaryTransformation):
+            return expression
+        raise exceptions.GraphQLNotSupportedByBackendError(
+            u'has_edge_degree filter operation is unsupported by SQL backend.'
+        )
+
+    new_ir_blocks = [
+        block.visit_and_update_expressions(visitor_fn)
+        for block in ir_blocks
+    ]
+    return new_ir_blocks
+
+
+def lower_optional_fields(ir_blocks, block_index_to_location, query_path_to_location_info):
+    """Lowering step for Filter blocks in an optional scope."""
+    def rewrite_optional(expression):
+        """Rewrite optional predicates to support the compiler's semantics.
+
+        For comparison to a field that originates from an optional context, the field will
+        be associated with a table that has been joined to the query with a LEFT JOIN.
+        The spec says that a predicate applied to this field should pass if this field doesn't
+        exist, which can be expressed in SQL as an (field IS NULL OR predicate) statement,
+        which can be constructed with this particular rewrite.
+        """
+        if not isinstance(expression, expressions.BinaryComposition):
+            return expression
+        if not isinstance(expression.left, expressions.LocalField):
+            return expression
+        if not isinstance(expression.right, (expressions.Variable, expressions.ContextField)):
+            return expression
+        # Either the expression is true, or
+        return expressions.BinaryComposition(
+            u'||',
+            # either the expression is true
+            expression,
+            # or the field referenced by the expression is NULL (doesn't exist)
```

Comment:
How do we know that the `left` is the field in question, and not the `right`? Seems like there should be an assertion here.